ShopperInsights/src/main/java/com/braintreepayments/api/shopperinsights/v2/CreateCustomerSessionApi.kt

```diff
@@ -0,0 +1,111 @@
+package com.braintreepayments.api.shopperinsights.v2
+
+import com.braintreepayments.api.core.BraintreeClient
+import com.braintreepayments.api.core.ExperimentalBetaApi
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * API to create a new customer session using the `CreateCustomerSession` GraphQL mutation.
+ */
+@ExperimentalBetaApi
+internal class CreateCustomerSessionApi(
+    private val braintreeClient: BraintreeClient,
+) {
+
+    sealed class CreateCustomerSessionResult {
+        data class Success(val sessionId: String) : CreateCustomerSessionResult()
+        data class Error(val error: Exception) : CreateCustomerSessionResult()
+    }
+
+    fun execute(
+        customerSessionRequest: CustomerSessionRequest,
+        callback: (CreateCustomerSessionResult) -> Unit
+    ) {
+        try {
+            val params = JSONObject()
+            params.put(
+                QUERY, """
+                mutation CreateCustomerSession(${'$'}input: CreateCustomerSessionInput!) {
+                    createCustomerSession(input: ${'$'}input) {
+                        sessionId
+                    }
+                }
+                """.trimIndent()
+            )
+
+            params.put(VARIABLES, assembleVariables(customerSessionRequest))
+
+            braintreeClient.sendGraphQLPOST(params) { responseBody: String?, httpError: Exception? ->
+                if (responseBody != null) {
+                    callback(CreateCustomerSessionResult.Success(parseSessionId(responseBody)))
+                } else if (httpError != null) {
+                    callback(CreateCustomerSessionResult.Error(httpError))
+                }
+            }
+        } catch (e: JSONException) {
+            callback(CreateCustomerSessionResult.Error(e))
+        }
+    }
+
+    @Throws(JSONException::class)
+    private fun assembleVariables(customerSessionRequest: CustomerSessionRequest): JSONObject {
+        val customer = JSONObject().apply {
+            putOpt(HASHED_EMAIL, customerSessionRequest.hashedEmail)
+            putOpt(HASHED_PHONE_NUMBER, customerSessionRequest.hashedPhoneNumber)
+            putOpt(PAYPAL_APP_INSTALLED, customerSessionRequest.payPalAppInstalled)
+            putOpt(VENMO_APP_INSTALLED, customerSessionRequest.venmoAppInstalled)
+        }
+
+        val purchaseUnits = customerSessionRequest.purchaseUnits
+            ?.takeIf { it.isNotEmpty() }
```

Comment:
Nice! I have not seen us use `takeIf` before. Seems really helpful for code like this vs a conditional null check. 🚀